editors/code/src/installation/language_server.ts

```diff
@@ -0,0 +1,119 @@
+import { unwrapNotNil } from "ts-not-nil";
+import { spawnSync } from "child_process";
+import * as vscode from "vscode";
+import * as path from "path";
+import { strict as assert } from "assert";
+import { promises as fs } from "fs";
+
+import { BinarySource, BinarySourceType, GithubBinarySource } from "./interfaces";
+import { fetchLatestArtifactMetadata } from "./fetch_latest_artifact_metadata";
+import { downloadFile } from "./download_file";
+
+export async function downloadLatestLanguageServer(
+    {file: artifactFileName, dir: installationDir, repo}: GithubBinarySource
+) {
+    const binaryMetadata = await fetchLatestArtifactMetadata({ artifactFileName, repo });
+
+    const {
+        releaseName,
+        downloadUrl
+    } = unwrapNotNil(binaryMetadata, `Latest GitHub release lacks "${artifactFileName}" file`);
+
+    await fs.mkdir(installationDir).catch(err => assert.strictEqual(
+        err && err.code,
+        "EEXIST",
+        `Couldn't create directory "${installationDir}" to download `+
+        `language server binary: ${err.message}`
+    ));
+
+    const installationPath = path.join(installationDir, artifactFileName);
+
+    await vscode.window.withProgress(
+        {
+            location: vscode.ProgressLocation.Notification,
+            cancellable: false, // FIXME: add support for canceling download?
+            title: `Downloading language server ${releaseName}`
+        },
+        async (progress, _) => {
+            let lastPrecentage = 0;
+            await downloadFile(downloadUrl, installationPath, (readBytes, totalBytes) => {
+                const newPercentage = (readBytes / totalBytes) * 100;
+                progress.report({
+                    message: newPercentage.toFixed(0) + "%",
+                    increment: newPercentage - lastPrecentage
+                });
+
+                lastPrecentage = newPercentage;
+            });
+        }
+    );
+
+    await fs.chmod(installationPath, 111); // Set xxx permissions
```

Comment:
`0o755` would be more in line with my expectations. `111` is.. `0o157`?

Reply:
Yes, sure I just thought to use least privilege here, bit I don't think it will matter match, 755 will be much simpler

Reply:
Well, `0x157` is read/write/execute for other users (not the owner), which is harmless but still.. At least this will fix you permission denied errors.